### ISSUE TITLE
small bug in SIP Via header generation

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2194]
|[<img alt="pitrou's avatar" src="https://avatars.githubusercontent.com/u/1721820?s=50" width="50" height="50">](https://github.com/pitrou)| @pitrou reported|
|-|-|
|Trac ID|trac#2194|
|Type|defect|
|Created|2006-10-26 10:56:12Z|
|Branch|https://github.com/twisted/twisted/tree/via-parameters-2194-2|

There is a small bug in the Via.toString() method. Perhaps it can only occur in theoretical conditions (I don't know enough about the SIP protocol to know for sure).

However, here it is:
```
>>> from twisted.protocols import sip
>>> via = sip.Via("localhost", ttl=2)
>>> via.toString()
'SIP/2.0/UDP localhost:5060;ttl=2'
>>> via = sip.Via("localhost", ttl=1)
>>> via.toString()
'SIP/2.0/UDP localhost:5060;ttl'
```

As you see, if a value is exactly 1, it isn't output. This is because of the test "```if value == True```" at line 207. Replacing it with "```if value is True```" fixes the bug.

<details><summary>Searchable metadata</summary>

```
trac-id__2194 2194
type__defect defect
reporter__antoine antoine
priority__highest highest
milestone__ 
branch__branches_via_parameters_2194_2 branches/via-parameters-2194-2
branch_author__washort washort
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1161860172000000 1161860172000000
changetime__1234976822000000 1234976822000000
version__None None
owner__washort washort

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="pitrou's avatar" src="https://avatars.githubusercontent.com/u/1721820?s=50" width="50" height="50">](https://github.com/pitrou)<a name="note_1"></a>|@pitrou commented|
|-|-|

And here is a test to be added in test_sip:ViaTestCase.

```
    def testOneIsNotTrue(self):
        s = "SIP/2.0/UDP example.com:5060;ttl=1"
        v = sip.parseViaHeader(s)
        self.assertEquals(v.toString().rsplit(';')[1], "ttl=1")
        self.checkRoundtrip(v)
```

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_2"></a>|@therve commented|
|-|-|

AFAIK, 1 is the default value for ttl.

**Comment by twisted-trac** [comment]
|[<img alt="pitrou's avatar" src="https://avatars.githubusercontent.com/u/1721820?s=50" width="50" height="50">](https://github.com/pitrou)<a name="note_3"></a>|@pitrou commented|
|-|-|

The bug can happen with any parameter, I just took "ttl" as an example.
Also, RFC 3261 says (section 18.2.2):
  If the address is a multicast address, the response SHOULD be sent using the TTL indicated in the "ttl" parameter, or with a TTL of 1 if that parameter is not present.

It does not say there is a default value if the parameter is present but does not specify a value.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to @washort|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="washort's avatar" src="https://avatars.githubusercontent.com/u/947487?s=50" width="50" height="50">](https://github.com/washort)<a name="note_5"></a>|@washort commented|
|-|-|

(In [25676]) Branching to 'via-parameters-2194'

**Comment by twisted-trac** [comment]
|[<img alt="washort's avatar" src="https://avatars.githubusercontent.com/u/947487?s=50" width="50" height="50">](https://github.com/washort)|@washort removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to @exarkun<br>@exarkun set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun set owner to @washort<br>@exarkun set status to `new`|
|-|-|

  1. twisted/protocols/sip.py
    1. In the class docstring for `Via` several attributes are documented as _Optional_ or _Required_.  I see that this is information from the spec, but I'm not sure what it means for the attribute.  This should be converted into information which can more easily be used by a developer likely to interact with this class (for example, perhaps the optional attributes should be documented as possibly having a `None` value if the information isn't present).
    1. The types of most of the attributes are not documented but should be.
    1. The enormous try/except in `parseViaHeader` doesn't seem ideal to me.  One specific problem it has is that some of the parsing code raises `SIPError` with detailed information about how the parsing failed (the wrong version number case), but the except suite squashes everything into a generic SIPError.  Another specific problem is the `log.err` call in the except suite; low-level parsing functions shouldn't log errors, they should raise exceptions indicating how parsing failed.  Another problem is that there seems to be no test coverage for the exception handling in this function; it's hard to really decide if the implementation strategy is appropriate without tests. :)
  1. The test suite hangs at `testAmoralRPort` for me.  A parse error shows up in the logs.  This may indicate that the Via rport handling has changed incompatibly...
  1. Why was `testExtraWhitespace` deleted?

**Comment by twisted-trac** [comment]
|[<img alt="washort's avatar" src="https://avatars.githubusercontent.com/u/947487?s=50" width="50" height="50">](https://github.com/washort)<a name="note_9"></a>|@washort commented|
|-|-|

(In [26100]) Branching to 'via-parameters-2194-2'

**Comment by twisted-trac** [comment]
|[<img alt="washort's avatar" src="https://avatars.githubusercontent.com/u/947487?s=50" width="50" height="50">](https://github.com/washort)<a name="note_10"></a>|@washort removed owner|
|-|-|

Fixed. The try/except turned out to be useless since nothing catches any exceptions at the call sites for parseViaHeader.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to @exarkun<br>@exarkun set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_12"></a>|@exarkun set owner to dash<br>@exarkun set status to `new`|
|-|-|

  1. It seems that there's still an incompatibility in the `rport` change.  Old code passing `True` will get inexpected behavior (_rport=True_, I think).  Can you deprecate this use of the `Via.__init__` `rport` parameter and restore the old behavior for that case?
  1. Before `testExtraWhitespace` was deleted.  Now it's duplicated.  Er?
  1. I don't think `_Absent.__nonzero__` quite does the trick.  Even if the deprecated code isn't being edited, its behavior is potentially being changed.  That's fine, if it has tests, but it doesn't appear to have tests.  So it probably needs to have tests added.  It also looks like some of it is broken by this change - for example, `Base._fixupNAT` compares the value of `rport` to `True`.  That case never happens anymore, right?  It's now the `None` case.

**Comment by twisted-trac** [comment]
|[<img alt="washort's avatar" src="https://avatars.githubusercontent.com/u/947487?s=50" width="50" height="50">](https://github.com/washort)<a name="note_13"></a>|@washort removed owner|
|-|-|

OK, new scheme for compatibility added (and other stuff fixed).

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_14"></a>|@glyph set owner to @washort|
|-|-|

  1. flakes: Not your fault, but in `test_sip`, twisted.cred.checkers and twisted.cred.portal are apparently unused.  Also, `twisted.cred.error` is imported in a weird way.
  1. `getWarningMethod()` should really be deprecated.  It's no longer necessary. just use `warnings.warn`.
  1. `BAD_REQUEST` is apparently never used.  Neither is a "400" used anywhere in the diff.  I guess it should just be deleted?
  1. I think it would be clearer to call the getter `_getrport`, rather than the ambiguous `rport`.  (Also the setter should be `_setrport`, since I find the case difference is a bit confusing.)
    1. also please two blank lines after the `rport` method, when you change its name.
  1. Given that `rportRequested` is a boolean, the "== True" crud is just noise.  Delete it.
  1. Docstrings should use `L{}` more often.  For example, instead of using a sentence fragment, `Via`'s docstring could begin with: "A L{Via} is a SIP 'Via' header...".  `parseViaHeader` should say "Parse a 'via' header, returning a L{Via}."  Also it could say "@return" and "@rtype".

This is all pretty cosmetic, so please address and then merge.

**Comment by twisted-trac** [comment]
|[<img alt="washort's avatar" src="https://avatars.githubusercontent.com/u/947487?s=50" width="50" height="50">](https://github.com/washort)<a name="note_15"></a>|@washort set status to `closed`|
|-|-|

(In [26236]) Merge via-parameters-2194-2.
Fix a defect in SIP Via header parameter generation.

Author: washort
Reviewers: exarkun, glyph
Closes [#2194](https://github.com/twisted/twisted/issues/2194).
